pkg/file/joiner/joiner.go

```diff
@@ -213,37 +213,36 @@ func (j *joiner) Seek(offset int64, whence int) (int64, error) {
 
 func (j *joiner) IterateChunkAddresses(fn swarm.AddressIterFunc) error {
 	// report root address
-	stop := fn(j.addr)
-	if stop {
-		return nil
+	err := fn(j.addr)
+	if err != nil {
+		return err
 	}
 
-	var eg errgroup.Group
-	j.processChunkAddresses(fn, j.rootData, j.span, &eg)
-
-	return eg.Wait()
+	return j.processChunkAddresses(j.ctx, fn, j.rootData, j.span)
 }
 
-func (j *joiner) processChunkAddresses(fn swarm.AddressIterFunc, data []byte, subTrieSize int64, eg *errgroup.Group) {
+func (j *joiner) processChunkAddresses(ctx context.Context, fn swarm.AddressIterFunc, data []byte, subTrieSize int64) error {
 	// we are at a leaf data chunk
 	if subTrieSize <= int64(len(data)) {
-		return
+		return nil
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
 	}
 
+	eg, ectx := errgroup.WithContext(ctx)
+
```

Comment:
No really, no. However I did have discussion with @acud, and decision was to leave the WaitGroup for now because it currently makes code more linear, and easier to follow (and I agree with decision). Maybe once we have some additional tests we can make the switch.